src/pyhf/optimize/opt_pytorch.py

```diff
@@ -1,58 +1,46 @@
-import torch.optim
-
-
-class pytorch_optimizer(object):
-    def __init__(self, **kwargs):
-        self.tensorlib = kwargs['tensorlib']
-        self.maxdelta = kwargs.get('maxdelta', 1e-5)
-        self.maxiter = kwargs.get('maxiter', 100000)
-
-    def unconstrained_bestfit(self, objective, data, pdf, init_pars, par_bounds):
-        init_pars = self.tensorlib.astensor(init_pars)
-        init_pars.requires_grad = True
-        optimizer = torch.optim.Adam([init_pars])
-        maxdelta = None
-        for i in range(self.maxiter):
-            loss = objective(init_pars, data, pdf)
-            optimizer.zero_grad()
-            loss.backward(retain_graph=True)
-            init_old = init_pars.data.clone()
-            optimizer.step()
-            maxdelta = (init_pars.data - init_old).abs().max()
-            if maxdelta < self.maxdelta:
-                break
-        return init_pars
-
-    def constrained_bestfit(
-        self, objective, constrained_mu, data, pdf, init_pars, par_bounds
-    ):
-        allvars = [
-            self.tensorlib.astensor(
-                [v] if i != pdf.config.poi_index else [constrained_mu]
-            )
-            for i, v in enumerate(init_pars)
-        ]
-        nuis_pars = [v for i, v in enumerate(allvars) if i != pdf.config.poi_index]
-        for np in nuis_pars:
-            np.requires_grad = True
-        poi_par = [v for i, v in enumerate(allvars) if i == pdf.config.poi_index][0]
-
-        def assemble(poi_par, nuis_pars):
-            pars = [x for x in nuis_pars]
-            pars.insert(pdf.config.poi_index, poi_par)
-            pars = self.tensorlib.concatenate(pars)
-            return pars
-
-        optimizer = torch.optim.Adam(nuis_pars)
-        for i in range(self.maxiter):
-            pars = assemble(poi_par, nuis_pars)
-            loss = objective(pars, data, pdf)
-            optimizer.zero_grad()
-            loss.backward(retain_graph=True)
-            optimizer.step()
-
-            after_pars = assemble(poi_par, nuis_pars)
-            maxdelta = (after_pars.data - pars.data).abs().max()
-            if maxdelta < self.maxdelta:
-                break
-        return pars
+from .. import get_backend, default_backend
+from ..tensor.common import _TensorViewer
+from .autodiff import AutoDiffOptimizerMixin
+import torch
+
+
+class pytorch_optimizer(AutoDiffOptimizerMixin):
+    def __init__(self, *args, **kargs):
+        pass
+
+    def setup_unconstrained(self, objective, pdf, data, init_pars, par_bounds):
+        def func(pars):
+            tensorlib, _ = get_backend()
+            pars = tensorlib.astensor(pars)
```

Comment:
why not `pars = tensorlib.astensor(pars, requires_grad=True)`?

Reply:
```suggestion
            pars = tensorlib.astensor(pars, requires_grad=True)
```

Reply:
Just to check, the only reason that we need to explicitly cast as a tensor here is because these `pars` are being passed into `func` by the internals of

```
fitresult = scipy.optimize.minimize(func, init, method='SLSQP', jac=True, bounds=bounds)
```

which won't be using tensors, correct? We can't avoid having to do this cast?

Reply:
because scipy doesn't know anything about tensors other than numpy, so we just use the backedns to compute the logpdf -> cast to numpy, and scipy.optimize will return a new parameter vector (as numpy array) at the location where it wants to evaluate a likelihood nexxt